EasyCommands/BlockHandlers/TurretBlockHandlers.cs

```diff
@@ -0,0 +1,87 @@
+﻿using Sandbox.Game.EntityComponents;
+using Sandbox.ModAPI.Ingame;
+using Sandbox.ModAPI.Interfaces;
+using SpaceEngineers.Game.ModAPI.Ingame;
+using System.Collections.Generic;
+using System.Collections;
+using System.Linq;
+using System.Text;
+using System;
+using VRage.Collections;
+using VRage.Game.Components;
+using VRage.Game.GUI.TextPanel;
+using VRage.Game.ModAPI.Ingame.Utilities;
+using VRage.Game.ModAPI.Ingame;
+using VRage.Game.ObjectBuilders.Definitions;
+using VRage.Game;
+using VRage;
+using VRageMath;
+
+namespace IngameScript {
+    partial class Program {
+        public class TurretControlBlockHandler : FunctionalBlockHandler<IMyTurretControlBlock> {
+            public TurretControlBlockHandler() {
+                AddNumericHandler(Property.RANGE, b => b.Range, (b, v) => b.Range = v, 100f);
+                AddPropertyHandler(Property.LOCKED, TerminalPropertyHandler("EnableTargetLocking", true));
+                AddPropertyHandler(Property.AUTO, TerminalPropertyHandler("AI", true));
+                AddBooleanHandler(Property.USE, b => b.IsUnderControl);
+                AddReturnHandlers(Property.TARGET, Return.VECTOR,
+                    TypeHandler(VectorHandler(b => {
+                        Vector3D? target = Vector3D.Zero;
+                        if (b.HasTarget) target = GetPosition(b.GetTargetedEntity());
+                        return target.Value;
+                    }), Return.VECTOR),
+                    TypeHandler(BooleanHandler(b => b.HasTarget), Return.BOOLEAN),
+                    TypeHandler(StringHandler(b => b.GetTargetingGroup(), (b, v) => b.SetTargetingGroup(v)), Return.STRING));
+                AddPropertyHandler(Property.ANGLE, TerminalPropertyHandler("AngleDeviation", 5f));
+                AddBooleanHandler(Property.TRIGGER, b => GetGuns(b).Any(g => g.IsShooting), (b, v) => {
+                    foreach (IMyUserControllableGun gun in GetGuns(b)) gun.Shoot = v;
+                });
+                AddDirectionHandlers(Property.VELOCITY, Direction.RIGHT,
+                    TypeHandler(NumericHandler(b => b.VelocityMultiplierAzimuthRpm, (b,v) => b.VelocityMultiplierAzimuthRpm = v), Direction.LEFT, Direction.RIGHT),
+                    TypeHandler(NumericHandler(b => b.VelocityMultiplierElevationRpm, (b,v) => b.VelocityMultiplierElevationRpm = v), Direction.UP, Direction.DOWN));
+                defaultPropertiesByPrimitive[Return.NUMERIC] = Property.RANGE;
+            }
+
+            IEnumerable<IMyUserControllableGun> GetGuns(IMyTurretControlBlock block) {
+                var tools = NewList<IMyFunctionalBlock>();
+                block.GetTools(tools);
+                return tools.Where(t => t is IMyUserControllableGun).Select(t => (IMyUserControllableGun)t);
```

Comment:
The Pattern `.Where(t => t is T).Select(t -> (T)t)` can always be replaced by `OfType<T>()`

Reply:
good call :)